### PR TITLE
Add check if tag contains indicator for pre release

### PR DESCRIPTION
**Pull request by soupdiver**
This PR addresses #667 
It's not complete yet but with a draft for the implementation.

The changes enable `goreleaser` to automatically mark a release as a pre release when the tag contains an indicator for it. A tag that would lead to a pre release would be `0.0.1-rc1`. The requirement is basically that `-` followed by 1+ characters is present.
This can probably be tweaked and hardened a bit. I'm open for suggestions!

I'm also not 100% sure yet if this is the best location for the code.
`ctx.Config.Release.Prerelease` is not accessed anywhere else so the choice where to put the code is basically free. But since we change the datatype from `bool` to `string` and `Context` does not have a member for `Prerelease` we would have to carry the variable around otherwise.

So either we check quite late in the stack (as I did) or we could add a Member to `Context` and have the check after reading the config file. I think the latter would be a better way but to demo my idea I implemented the first choice.

Todo:
- [x] Add tests
- [x] Clarify when to check for pre release in tag

### CODE REVIEW

**Comment by caarlos0** [comment]
we already use the masterminds/semver lib, and it already has a method to find out if the a version is a prerelease: https://godoc.org/github.com/Masterminds/semver#Version.Prerelease

we are already parsing the version in the git pipe I think... maybe we can make this even simpler?

**Comment by caarlos0** [comment]
oh, and thanks for the PR :D

**Comment by soupdiver** [comment]
@caarlos0 Simplified the code. What do you think?

**Comment by caarlos0** [comment]
Looks good, but i think a better place for is in the release pipe... 

We could also have a `PreRelease` field on `context.Git`, so we can parse the tag a single time (maybe on the git pipe)...

What do you think?

**Comment by soupdiver** [comment]
I think I found a good place where to put the logic and also added a test.
Not sure about codecov results.

File: internal/client/github.go [hunk 1]
```diff
@@ -94,12 +94,28 @@ func (c *githubClient) CreateRelease(ctx *context.Context, body string) (int64,
 	if err != nil {
 		return 0, err
 	}
+
+	preRelease := false
+	// Check if we have to check the git tag for an indicator to mark as pre release
+	switch ctx.Config.Release.Prerelease {
```
Comment (caarlos0) [inline internal/client/github.go]: i would move this part to the release pipe itself... what do you think?

**Comment by caarlos0** [comment]
Hi @soupdiver 

one small comment, and after doing it we can probably increase the test coverage again :D 

Thanks again for the PR and all the effort 🚀

**Comment by soupdiver** [comment]
@caarlos0 I had a look at your suggestion. I put it into `github.go` because I thought about the case if some hoster does not provide pre-releases. That's why I put the logic into the specific implementation.
If we only support github this maybe doesn't make _that_ much of a difference.

Another thing:
Since `ctx.Config.Release.Prerelease` holds a string value (true, false or auto) we have to map this to a boolean at some point. I could put the logic of parsing and decision if we mark as pre-release into `release.go`. But then I would have to add another variable to context or release to hold this boolean. Otherwise I could add the boolean as a parameter to `CreateRelease` which would change the interface and cause need for changes in other places as well I think.

Both are possible. Which one do you prefer?

**Comment by caarlos0** [comment]
hmm, and there is another "problem", maybe, at some point, users will want to use the pre-release part of the tag on name templates... I guess we could either parse the tag multiple times or put all this in a single place...  anyway, we can leave this for later.

about the best place to put the decision, the main argument to put it in the release pipe `Default()` is that there is a pipe that writes the "effective config file" to dist (which I usually ask for on bug issues), so we will know before actually releasing what kind of release it will do... 

about other release providers that may not support pre-releases, I think we would just not pass that option to them in that case... anyway, it is a speculation, as now we don't have others haha 

all that being said, I would put that logic on the `release.Default` 🤔
?

**Comment by soupdiver** [comment]
Can I expected that `ctx.Git` is already populated when `pipe.Default` is run?

**Comment by caarlos0** [comment]
> Can I expected that `ctx.Git` is already populated when `pipe.Default` is run?

hmm, right now, no, but we should... let me fix that!

**Comment by caarlos0** [comment]
#879

**Comment by soupdiver** [comment]
@caarlos0 I took the existing tests as a template for mine but something goes wrong 🤔 
It seems the values set at
https://github.com/goreleaser/goreleaser/blob/0e0631a8d4616d4ec4c1ad66a62425ce5cfb273c/internal/pipe/release/release_test.go#L169
and
https://github.com/goreleaser/goreleaser/blob/0e0631a8d4616d4ec4c1ad66a62425ce5cfb273c/internal/pipe/release/release_test.go#L173

are not passed down the pipeline somehow.

See: https://travis-ci.org/goreleaser/goreleaser/builds/460234527#L712
and https://travis-ci.org/goreleaser/goreleaser/builds/460234527#L718

you have an idea what's wrong there?

**Comment by caarlos0** [comment]
I'll try to take a look later...

**Comment by caarlos0** [comment]
Oh, inside the release tests, the pipe that detects the git tag does not run, that's why it is not working.

you can cherry-pick this commit https://github.com/goreleaser/goreleaser/commit/fb8e5701814c3f7fbe23b5d7a60640d3cb74f46b and it will be fixed.

**Comment by soupdiver** [comment]
check, will try

**Pull request merged**